Tests/TensorFlowTests/EpochsTests.swift

```diff
@@ -0,0 +1,307 @@
+// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import XCTest
+
+@testable import TensorFlow
+
+var rng = ARC4RandomNumberGenerator(seed: [42])
+
+final class EpochsTests: XCTestCase {
+  func testBaseUse() {
+    // A mock item type that tracks if it was accessed or not
+    class Tracker {
+      var accessed: Bool = false
+    }
+
+    // `inBatches` splits our dataset in batches, the `collated` property is
+    // defined for any struct conforming to `Collatable`
+    let rawItems = Array(0..<512).map { _ in Tracker() }
+    let dataset = rawItems.lazy.map { (x: Tracker) -> Tensor<Float> in
+      x.accessed = true
+      // Using a random tensor here is not thread-safe and will result in race
+      // conditions.
+      return Tensor<Float>(randomNormal: [224, 224, 3])
+    }
+    let batches = dataset.inBatches(of: 64).lazy.map(\.collated)
+
+    for (i, batch) in batches.enumerated() {
+      XCTAssertEqual(batch.shape, TensorShape([64, 224, 224, 3]))
+      let limit = (i + 1) * 64
+      XCTAssert(rawItems[..<limit].allSatisfy(\.accessed))
+      XCTAssert(rawItems[limit...].allSatisfy({ !$0.accessed }))
+    }
+  }
+
+  // Tests with shuffle
+  func testShuffle() {
+    // Using `dataset.shuffled()` would break the laziness. Plus we would need
+    // to do it at each new epoch. `TrainingEpochs` automatically handles
+    // shuffling (and re-shuffling at each epoch) without breaking the laziness.
+    let dataset = (0..<512).lazy.map { (i: Int32) -> Tensor<Int32> in
+      return Tensor<Int32>(zeros: [32, 32, 3]) + i
+    }
+
+    let epochs = TrainingEpochs(samples: dataset, batchSize: 64, entropy: rng)
+    var accessed = Array(0..<512)
+    for batches in epochs.prefix(10) {
+      var newAccessed: [Int] = []
+      for batch in batches {
+        XCTAssertEqual(batches.count, 8)
+        let collatedBatch = batch.collated
+        XCTAssertEqual(collatedBatch.shape, TensorShape([64, 32, 32, 3]))
+
+        newAccessed += Array(0..<64).map {
+          Int(collatedBatch[$0, 0, 0, 0].scalarized())
+        }
+      }
+      XCTAssertNotEqual(
+        accessed, newAccessed,
+        "Dataset should have been reshuffled.")
+
+      accessed = newAccessed
+      let uniqueSamples = Set(accessed)
+      XCTAssertEqual(
+        uniqueSamples.count, 512,
```

Comment:
Avoid using a magic number that needs to be kept up-to-date with other parts of the code.

```suggestion
        uniqueSamples.count, accessed.count,
```